tracer.go

```diff
@@ -222,22 +222,30 @@ func (t *Tracer) startSpanWithOptions(
 	var references []Reference
 	var parent SpanContext
 	var hasParent bool // need this because `parent` is a value, not reference
+	var ctx SpanContext
+	var isSelfRef bool
 	for _, ref := range options.References {
-		ctx, ok := ref.ReferencedContext.(SpanContext)
+		ctxRef, ok := ref.ReferencedContext.(SpanContext)
 		if !ok {
 			t.logger.Error(fmt.Sprintf(
 				"Reference contains invalid type of SpanReference: %s",
 				reflect.ValueOf(ref.ReferencedContext)))
 			continue
 		}
-		if !isValidReference(ctx) {
+		if !isValidReference(ctxRef) {
 			continue
 		}
-		references = append(references, Reference{Type: ref.Type, Context: ctx})
+		references = append(references, Reference{Type: ref.Type, Context: ctxRef})
+
 		if !hasParent {
-			parent = ctx
+			parent = ctxRef
 			hasParent = ref.Type == opentracing.ChildOfRef
 		}
+
+		if ref.Type == SelfRef {
```

Comment:
I think you want this check earlier, before calling append() in L238, because otherwise `references` list will contain a reference pointing to self, which may confuse the backend.